WIP: Backports for 1.4-RC1

KristofferC opened the pull request:
Backported PRs:
- [x] #34237 - outline length assert in default_rng(tid)
- [x] #34223 - fix #33954, recursion through field types in `is_derived_type`
- [x] #34243 - fix codegen for getfield of homogeneous tuples
- [x] #34230 - fix #32442, broadcasting over non-offset arrays with mismatched axis eltypes
- [x] #30239 - Update svd.jl
- [x] #34299 - Fix trailing whitespace in svd docs introduced in cec4c3
- [x] #34282 - add `local` to testset vars that should be local
- [x] #34291 - fix #34286, regression in `methods` with empty tuple of types
- [x] #34287 - make fallback `signed` and `unsigned` functions unchecked (as documented)
- [x] #34289 - add some compatibility shims for internal functions changed in 1.4
- [x] #34284 - add a compatibility fallback for `divrem` and `fldmod`
- [x] #34303 - Simplify the GC.gc interface
- [x] #34319 - Update Mozilla CA certificate store to latest (01-01-2020) for libgit2 SSL.
- [x] #34314 - restore `pathsep()` and old sig of `print_matrix_vdots` for backwards compat
- [x] #34322 - handle `GlobalRef` in getfield elim pass
- [x] #34257 - Update sprandn in docs.
- [x] #34316 - add `hypot` fallback for 2 Numbers of the same type
- [x] #34361 - Use approximate comparison in LinearAlgebra diagonal tests.
- [x] #34326 - Generate an error for closure cfunctions on unsupported platforms.
- [x] #34369 - Fix initial value handling with flatten transducer
- [x] #34379 - Change lowering of gc preserve
- [x] #34349 - Force full collection in SharedArrays tests
- [x] #34225 - Fix CI hangs on win32 
- [x] #34129  - Make it easier to extend lcm/gcd/gcdx to custom types
- [x] #33785 - Use UUID's to generate random tempname on Windows 
- [x] #34429 - Check return value of SystemFunction036 
- [x] #34405 - remove code forcing testsets to be compiled 
- [x] #34433-  fix #34415, remove affinity setting code (

@nanosoldier `runbenchmarks(ALL, vs = ":release-1.3")`

sostock committed:
Make it easier to extend lcm/gcd/gcdx to custom types

(cherry picked from commit 85eb20f9cab2100c3688029ab5123c366432b20e)

Keno committed:
Fix CI hangs on win32

After much debugging, it turns out that OpenBlas is sometimes hanging
on exit due to bad assumptions in its thread management. The upstream
discussion is at https://github.com/xianyi/OpenBLAS/pull/2350, but this
should fix our frequent win32 CI failures in the meantime.

KristofferC committed:
outline rng length assert

(cherry picked from commit d723ceefeba39d12d874a124aa8e37014d1b44ec)

KristofferC commented:
I had the wrong Base branch when starting nanosoldiers but I don't think anything has really been merged so should be ok.

KristofferC commented:
@nanosoldier `runtests(ALL, vs = ":release-1.3")`

JeffBezanson committed:
fix #33954, recursion through field types in `is_derived_type` (#34223)

(cherry picked from commit 8b57f642c45b6158db0dce145e1c1aa8c1ba425c)

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/34238) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/18193e50f19a372d877186ffe82683288e13a866/bb47292_vs_2e6715c/report.md). cc @ararslan

nanosoldier commented:
[Your test job](https://github.com/JuliaLang/julia/pull/34238#issuecomment-570281523) has completed - possible new issues were detected. A full report can be found [here](https://github.com/maleadt/BasePkgEvalReports/blob/0bb12607c245153c8e7215788e580d43d7d2d51a/pkgeval-9bd498d_vs_2e6715c/report.md). cc @maleadt

KristofferC commented:
Some failures from https://github.com/JuliaLang/Pkg.jl/pull/1548 not being included here.

JeffBezanson committed:
fix codegen for getfield of homogeneous tuples (#34243)

fix #34206, fix #34207

(cherry picked from commit 007cb29e9d435026df87c330f9fc0df9907c9525)

mbauman committed:
fix #32442, broadcasting over non-offset arrays with mismatched axis eltypes (#34230)

In cases where we have multiple arrays with `OneTo` axes that do not share the same axis eltype, we should simply default to constructing a new array with `OneTo{Int}` axes.

(cherry picked from commit 11e7c3377163fb730b81b0e782e37f5958ffa747)

JeffBezanson commented:
AIBECS: BSON
ARCHModels: numerical issue
AbstractNumbers: fails because `div` is now `applicable` to Complex
  even though it doesn't work
AdaptiveDistanceFields: AxisAlgorithms: bad Vararg length; looks fixed (upgrade a dependency?)
Atom: pathsep not defined
BSON: https://github.com/JuliaIO/BSON.jl/pull/63
BandedMatrices: no method matching bunchkaufman!(::Symmetric{Float32,BandedMatrix{Float32,Array{Float32,2},Base.OneTo{Int64}}}, ::Bool; check=false)

JeffBezanson commented:
BitIntegers: #33459. we removed a fallback `signed()` method it used.
CBinding: jl_gc_total_bytes renamed; use Base.gc_bytes() instead
CategoricalArrays: different whitespace in output
CatmullClark: #33954, fixed
CurrenciesBase: FixedDecimals: another div issue (#34284)
DBFTables: output changed
DataInterpolations: Zygote needs to handle loopinfo https://github.com/FluxML/Zygote.jl/issues/393
DecisionTree: code coverage?
DiffEqParamEstim: Zygote
DiffEqSensitivity: Zygote
Distributions: NaN === NaN ?
Econometrics: output changed
EfficientGlobalOptimization: Zygote
ExactPredicates: #34252
FilePathsBase: BSON
FixedEffectModels: PosDefException: matrix is not positive definite; Cholesky factorization failed.
Flux: Zygote
GFlops: some numbers don't match; not sure
GLM: PosDefException: matrix is not positive definite; Cholesky factorization failed.
GeoArrays: curl issue?
GeometricFlux: Zygote
GeometricIntegratorsDiffEq: DecFP
HalfIntegers: div.jl issue with cld; packages need to implement div(a, b, RoundUp)
Hecke: maybe due to changes in reduce?
   [1] _empty_reduce_error() at ./reduce.jl:292
   [2] reduce_empty(::Function, ::Type{T} where T) at ./reduce.jl:302
   [3] reduce_empty(::Base.BottomRF{typeof(max)}, ::Type{T} where T) at ./reduce.jl:317
   [4] mapreduce_empty(::typeof(abs), ::Base.BottomRF{typeof(max)}, ::Type{T} where T) at ./reduce.jl:333
HomotopyContinuation: `@allocated` changes
Hyperspecialize: crash
IncrementalInference: numerical issue?
InfrastructureSystems: #34243
JWAS: zero(Any) ?
JuMP: output changes
KernelFunctions: Zygote

KristofferC commented:
I started a bit (from the bottom) here: https://docs.google.com/spreadsheets/d/1ngalhKl9FEplWCWAWOfVKamZj-pkt4157pFGimb5wCw/edit?usp=sharing

vtjnash commented:
DecisionTree was a known bug in that package (since fixed)

Hyperspecialize is probably an instance of https://github.com/JuliaLang/julia/issues/34252 (prior to this release, `@testset` used to expand to heuristics that would force off inference and optimization. It has forced compilation heuristics since 057fa2aa9f9, we just would not infer the code first.)

alanedelman committed:
Update svd.jl (#30239)

* Update svd.jl

* Update svd.jl

Make doc more useful to the reader.

* Fix grammar and one trailing whitespace

* Update svd.jl

* Fix doctests

* Avoid duplicating documentation between mutating and non-mutating
versions of svd functions.

* Reorganize generalized svd docstring and doctests

* Break some long lines

Co-authored-by: Viral B. Shah <viral@juliacomputing.com>
Co-authored-by: Andreas Noack <andreas@noack.dk>
(cherry picked from commit cec4c3210986ba045f0d8ecf6396647337361d8e)

fredrikekre committed:
Fix trailing whitespace in svd docs introduced in cec4c3210986ba045f0d8ecf6396647337361d8e.

(cherry picked from commit a3eb2749b48aa1b650a103ad06e07074c0608ce2)

JeffBezanson committed:
add `local` to testset vars that should be local (#34282)

(cherry picked from commit 194a38bbe22168556d4d0bd816a947abeb5c2f0e)

JeffBezanson committed:
fix #34286, regression in `methods` with empty tuple of types (#34291)

(cherry picked from commit 8dc0d9332d1cbe204ae3ea53457988f3ea2029ee)

JeffBezanson committed:
make fallback `signed` and `unsigned` functions unchecked (as documented) (#34287)

(cherry picked from commit 37275c3f136b3a1b3bd6215432f021a22371d536)

JeffBezanson committed:
add some compatibility shims for internal functions changed in 1.4 (#34289)

(cherry picked from commit c33298df0b58a2a3d6a5a2e5afce6093061888c9)

JeffBezanson committed:
restrict `div` fallback to `Real` (#34284)

add more-compatible fallback for `divrem`

(cherry picked from commit 8f53987306e34d63da94dfaa66201aa009178e75)

JeffBezanson commented:
> prior to this release, `@testset` used to expand to heuristics that would force off inference and optimization

Do you happen to know what that change was?

vtjnash commented:
b9546c8b487364cd998b3ad87d873dc3ef35ac43

maleadt committed:
Simplify the GC.gc interface. (#34303)

(cherry picked from commit b0ed147c5cf0566019ff08755af71eb5649411bc)

sostock commented:
>  HalfIntegers: div.jl issue with cld; packages need to implement div(a, b, RoundUp)

This is [fixed on HalfIntegers master](https://github.com/sostock/HalfIntegers.jl/commit/0755b97be9c71a7f83a7ca79163a3e1ebcddf27c), but I wanted to wait for a decision on #34129 before releasing a new HalfIntegers version that contains the fix.

andreasnoack commented:
Whitespace error fixed in https://github.com/JuliaLang/julia/pull/34299

mikhail-j committed:
Update Mozilla CA certificate store to latest (01-01-2020) for libgit2 SSL. (#34319)

(cherry picked from commit 375347ea52b27ffc21b5e452b4d094074a15d86e)

staticfloat committed:
Bump OpenBLAS BB release to include win32 threading patch (#34273)

JeffBezanson committed:
restore `pathsep()` and old sig of `print_matrix_vdots` for backwards compat (#34314)

(cherry picked from commit e3b9d5df3b5e971de3608305e3aa0f8c23f6c5ee)

JeffBezanson committed:
handle `GlobalRef` in getfield elim pass (#34322)

(cherry picked from commit 22817ea28366b17a982c00d405edebe8c9571d5c)

KristofferC committed:
bump Pkg to release-1.4

maleadt commented:
@nanosoldier `runtests(ALL, vs = ":release-1.3")`

maleadt committed:
Generate an error for closure cfunctions on unsupported platforms.

(cherry picked from commit f53e144c76530fc164f8faebcc1dff16b7f05590)

ViralBShah committed:
Update sprandn in docs. Fix #31283 (#34257)

(cherry picked from commit 717d54894cb3a921c66509b86227db088e3bde33)

nanosoldier commented:
[Your test job](https://github.com/JuliaLang/julia/pull/34238#issuecomment-572547430) has completed - possible new issues were detected. A full report can be found [here](https://github.com/maleadt/BasePkgEvalReports/blob/f062a47c6bb3b5dd7c7bd29e7a3cc53cc6feeedd/pkgeval/by_hash/c32d166_vs_2e6715c/report.md). cc @maleadt

KristofferC commented:
221 to 131 failures is some kind of progress...

thofma commented:
> Hecke: maybe due to changes in reduce?
> [1] _empty_reduce_error() at ./reduce.jl:292
> [2] reduce_empty(::Function, ::Type{T} where T) at ./reduce.jl:302
> [3] reduce_empty(::Base.BottomRF{typeof(max)}, ::Type{T} where T) at ./reduce.jl:317
> [4] mapreduce_empty(::typeof(abs), ::Base.BottomRF{typeof(max)}, ::Type{T} where T) at ./reduce.jl:333

It is a bug in Hecke. Will be fixed soonish.

KristofferC commented:
Here is a categorization of the changes: https://docs.google.com/spreadsheets/d/1ngalhKl9FEplWCWAWOfVKamZj-pkt4157pFGimb5wCw/edit?usp=sharing

- Green is for things where I don't think we need to change anything in Julia
- Orange is when I am not sure
- Red is when we need to fix Julia or investigate further
- Yellow is when another package causes the failure
- Teal is when this is an inference change (like `@inferred` or counting allocations)

Keno committed:
Force full collection in SharedArrays tests (#34349)

Fixes the same CI failure as https://github.com/JuliaLang/julia/pull/34336,
but hopefully should be immediately mergeable without objection about what
the interface is.

(cherry picked from commit 3ed4e943616b45ed8be4216c2e0c52adc17bb6ba)

JeffBezanson committed:
add `hypot` fallback for 2 Numbers of the same type (#34316)

(cherry picked from commit ade471359849cb2767eed69921892ab4280c3b81)

KristofferC commented:
Pkg should have some fixes, Zygote should be updated, some packages might be fixed.

@nanosoldier `runtests(["ARCHModels", "AbstractNumbers", "AdaptiveDistanceFields", "AltDistributions", "ApproxManifoldProducts", "BSONqs", "BandedMatrices", "BitIntegers", "BufferedStreams", "CBinding", "COBRA", "COSMO", "ConicBenchmarkUtilities", "CuthillMcKee", "DataInterpolations", "DensityRatioEstimation", "DiffEqParamEstim", "DisplayAs", "Distributions", "DistributionsAD", "DynamicPPL", "Econometrics", "EfficientGlobalOptimization", "ExactPredicates", "EzXML", "FameSVD", "FlexibilityAnalysis", "Flux", "GARCH", "GFlops", "GLPKMathProgInterface", "Gadfly", "GeoArrays", "GeometricFlux", "GeometricIntegratorsDiffEq", "GlobalSearchRegressionGUI", "GoogleCloudObjectStores", "HDF5", "HalfIntegers", "Hecke", "HomotopyContinuation", "Hyperspecialize", "IncrementalInference", "InspectDR", "InteractiveCodeSearch", "JWAS", "JuMP", "Kaleido", "KernelFunctions", "KrigingModel", "LSL", "LanguageServer", "LazyArrays", "LazySets", "LazyStack", "LengthChannels", "LibSndFile", "Libtask", "LocalCoverage", "LogDensityProblems", "LuxurySparse", "MAT", "Mads", "Match", "MathOptInterface", "MathematicalSystems", "MbedTLS", "Measurements", "MemoryBasedCF", "Metalhead", "MicroLogging", "Microbiome", "Mill", "Missings", "MolecularTrajectories", "Munkres", "NLreg", "NaiveGAflux", "NaiveNASflux", "NamedDims", "NetworkInference", "OMEinsum", "OOESAlgorithm", "ObjectStores", "OceanTurb", "POMDPModelTools", "POMDPs", "Parametron", "PiCraft", "PkgBenchmark", "PkgDev", "Polyhedra", "PolynomialAmoebas", "PowerModelsACDC", "PushVectors", "PyCallJLD", "RandomBasedArrays", "RayTracer", "RegistryTools", "RetroCap", "RiemannTheta", "RoME", "RoundingIntegers", "SatelliteToolbox", "Sched", "SimpleDiffEq", "SimpleMock", "SliceMap", "SparseDiffTools", "SparsityDetection", "StaticArrays", "StaticUnivariatePolynomials", "Stheno", "Strided", "Surrogates", "SymbolServer", "TakagiFactorization", "TensorDecompositions", "TensorNetworkAD", "TimeZones", "ToStruct", "Transformers", "TuringModels", "TypeSortedCollections", "Unitful", "VQC", "VerTeX", "WebSockets", "YaoArrayRegister", "YaoBlocks", "Zygote"], vs = ":release-1.3")`

KristofferC commented:
@nanosoldier `runtests(["ARCHModels", "AbstractNumbers", "AdaptiveDistanceFields", "AltDistributions", "ApproxManifoldProducts", "BSONqs", "BandedMatrices", "BitIntegers", "BufferedStreams", "CBinding", "COBRA", "COSMO", "ConicBenchmarkUtilities", "CuthillMcKee", "DataInterpolations", "DensityRatioEstimation", "DiffEqParamEstim", "DisplayAs", "Distributions", "DistributionsAD", "DynamicPPL", "Econometrics", "EfficientGlobalOptimization", "ExactPredicates", "EzXML", "FameSVD", "FlexibilityAnalysis", "Flux", "GARCH", "GFlops", "GLPKMathProgInterface", "Gadfly", "GeoArrays", "GeometricFlux", "GeometricIntegratorsDiffEq", "GlobalSearchRegressionGUI", "GoogleCloudObjectStores", "HDF5", "HalfIntegers", "Hecke", "HomotopyContinuation", "Hyperspecialize", "IncrementalInference", "InspectDR", "InteractiveCodeSearch", "JWAS", "JuMP", "Kaleido", "KernelFunctions", "KrigingModel", "LSL", "LanguageServer", "LazyArrays", "LazySets", "LazyStack", "LengthChannels", "LibSndFile", "Libtask", "LocalCoverage", "LogDensityProblems", "LuxurySparse", "MAT", "Mads", "Match", "MathOptInterface", "MathematicalSystems", "MbedTLS", "Measurements", "MemoryBasedCF", "Metalhead", "MicroLogging", "Microbiome", "Mill", "Missings", "MolecularTrajectories", "Munkres", "NLreg", "NaiveGAflux", "NaiveNASflux", "NamedDims", "NetworkInference", "OMEinsum", "OOESAlgorithm", "ObjectStores", "OceanTurb", "POMDPModelTools", "POMDPs", "Parametron", "PiCraft", "PkgBenchmark", "PkgDev", "Polyhedra", "PolynomialAmoebas", "PowerModelsACDC", "PushVectors", "PyCallJLD", "RandomBasedArrays", "RayTracer", "RegistryTools", "RetroCap", "RiemannTheta", "RoME", "RoundingIntegers", "SatelliteToolbox", "Sched", "SimpleDiffEq", "SimpleMock", "SliceMap", "SparseDiffTools", "SparsityDetection", "StaticArrays", "StaticUnivariatePolynomials", "Stheno", "Strided", "Surrogates", "SymbolServer", "TakagiFactorization", "TensorDecompositions", "TensorNetworkAD", "TimeZones", "ToStruct", "Transformers", "TuringModels", "TypeSortedCollections", "Unitful", "VQC", "VerTeX", "WebSockets", "YaoArrayRegister", "YaoBlocks", "Zygote"], vs = ":release-1.3")`

maleadt committed:
Use approximate comparison in LinearAlgebra diagonal tests.

(cherry picked from commit 78d135257fa49fc2e0f160a70b7fe6fd476200aa)

nanosoldier commented:
[Your test job](https://github.com/JuliaLang/julia/pull/34238#issuecomment-573727761) has completed - possible new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/2bf4d9964d49d0812a694af248c3868f68fd49b6/pkgeval/by_hash/6bf2ec5_vs_2e6715c/report.md). cc @maleadt

KristofferC commented:
131 to 85...

KristofferC commented:
A summary of things we should probably fix / look at.

- AbstractNumbers: `no method matching div(::Complex{Float64}, ::Complex{Float64}`
- Distributions: `NaN === NaN`
- HalfIntegers: `No method matching div(::Half{Int64}, ::Half{Int64}, ::RoundingMode{:Down})`
- RoundingNumbers: `  MethodError: no method matching div(::RInt64, ::RInt64, ::RoundingMode{:ToZero})`
- AdaptiveDistanceFields: `AxisAlgorithms: ERROR: LoadError: LoadError: TypeVar in Vararg length must have bounds Union{} and Any`
- Gadfly: `No package 'libffi' found`
- JWAS: `zero(::Type{Any})` (sparse broadcasting depending on inference??)
- LazySets: `Rmath not properly installed. Please run Pkg.build("Rmath") and restart julia`
- LSL: Tests hang
- SimpleDiffEq: Tests hang
- BitIntegers: Assertion error
- Strided: Aborts (https://github.com/JuliaLang/julia/issues/34247)
- MbedTLS: Segfaults
- SymbolServer: Segfaults

KristofferC commented:
> - AdaptiveDistanceFields: bisected to #34091 (yes, I double checked)

AdaptiveDistanceFields is because different version of AxisAlgorithms end up getting installed.

JeffBezanson commented:
We used to have this `div` fallback:
```
div(x::T, y::T) where {T<:Real} = convert(T,round((x-rem(x,y))/y))
```
but it was removed (and similar for `fld`). If we put that back (in 3 argument form) it would probably fix those failures.

Keno commented:
> We used to have this `div` fallback

That fallback is very bad - it basically never gives a correct answer - a method error is much preferable.

tkf committed:
Fix initial value handling with flatten transducer (#34369)

There was a bug in initial value handling of `FlatteningRF` and the
following example failed:

    @test mapfoldl(
        x -> (x, x),
        ((a, b), (c, d)) -> (min(a, c), max(b, d)),
        Iterators.flatten((1:2, 3:4)),
    ) == (1, 4)

This is because `BottomRF(op.rf)` was called inside `FlatteningRF`
where `op.rf` is already a "non-bottom" reducing function; here it's a
`MappingRF`.  As `BottomRF(rf)` forwards anything on the second
argument on the first invocation as the first argument (accumulator)
of the next calls, we need to make sure that this value is processed
through `MappingRF` in the above example.  However, if we do
`BottomRF(op.rf)` where `op.rf` is a `MappingRF`, this `BottomRF`
bypasses any processing that has to happen in `op.rf`.

(cherry picked from commit 0ee3264be377a70e56465142649e67b4f85dda12)

tkf commented:
It'd be nice if https://github.com/JuliaLang/julia/pull/34369 is backported. It's a bug fix.

Keno committed:
Change lowering of gc preserve (#34379)

This fixes #34247 by changing the way gc preserve is lowered.
Instead of lowering it in a macro, lower it in the frontend.
This allows us to use an SSA value directly for the return token
of the gc begin expression. This bypasses the slot-renaming pass
of the compiler, thus preventing the compiler from trying to save
and restore the token. Of course, this kind of code would generally
not be legal (because it uses an SSA value outside of the regular
domination relation), but since this is a julia restriction, not
an LLVM restriction, we can simply exempt gc_begin tokens from this
particular validation. This works fine at the LLVM level also, because
it doesn't have this particular restriction. It also doesn't have
the same correctness problems as doing the same for non-token values,
as the tokens get lowered away by the try/catch lowering before reaching
the LLVM backend.

(cherry picked from commit 07a16d695feeba66db2a44c640675e7974ae870b)

maleadt commented:
Some more overnight bisection results:

- JWAS: Allow more general eltypes in sparse array multiplication (#33205)
- Polyhedra: fix some cases of `typeinfo` printing (#33290)
- Distributions: make allocated macro more reliable (#34091)
- SimpleDiffEQ: Upgrade to LLVM v8.0.1 (#32712)

KristofferC commented:
MbedTLS fails intermittently locally with this log

<details> <summary>Log </summary> 

```
Internal error: encountered unexpected error in runtime:
TypeError(func=:Tuple, context="parameter", expected=Type, got=<?#0x104ac61f0::<circular reference @-1>>)
jl_type_error_rt at /Users/sabae/buildbot/worker/package_macos64/build/src/rtutils.c:118
jl_f_apply_type at /Users/sabae/buildbot/worker/package_macos64/build/src/builtins.c:0
jl_apply at /Users/sabae/buildbot/worker/package_macos64/build/src/./julia.h:1692 [inlined]
do_apply at /Users/sabae/buildbot/worker/package_macos64/build/src/builtins.c:643
argtypes_to_type at ./compiler/typeutils.jl:46 [inlined]
abstract_call_known at ./compiler/abstractinterpretation.jl:872
abstract_call at ./compiler/abstractinterpretation.jl:895
abstract_call at ./compiler/abstractinterpretation.jl:880
abstract_eval at ./compiler/abstractinterpretation.jl:974
typeinf_local at ./compiler/abstractinterpretation.jl:1227
typeinf_nocycle at ./compiler/abstractinterpretation.jl:1283
typeinf at ./compiler/typeinfer.jl:12
typeinf_ext at ./compiler/typeinfer.jl:574
typeinf_ext at ./compiler/typeinfer.jl:605
jfptr_typeinf_ext_1.clone_1 at /Applications/Julia-1.5.app/Contents/Resources/julia/lib/julia/sys.dylib (unknown line)
jl_apply at /Users/sabae/buildbot/worker/package_macos64/build/src/./julia.h:1692 [inlined]
jl_type_infer at /Users/sabae/buildbot/worker/package_macos64/build/src/gf.c:213
jl_compile_method_internal at /Users/sabae/buildbot/worker/package_macos64/build/src/gf.c:1887
_jl_invoke at /Users/sabae/buildbot/worker/package_macos64/build/src/gf.c:2153 [inlined]
jl_apply_generic at /Users/sabae/buildbot/worker/package_macos64/build/src/gf.c:2322
unknown function (ip: 0x11f36af7c)
ssl_generate_random at /Users/kristoffer/.julia/packages/MbedTLS/a1JFn/deps/usr/lib/libmbedtls.2.16.0.dylib (unknown line)
ssl_write_client_hello at /Users/kristoffer/.julia/packages/MbedTLS/a1JFn/deps/usr/lib/libmbedtls.2.16.0.dylib (unknown line)
mbedtls_ssl_handshake_client_step at /Users/kristoffer/.julia/packages/MbedTLS/a1JFn/deps/usr/lib/libmbedtls.2.16.0.dylib (unknown line)
mbedtls_ssl_handshake_step at /Users/kristoffer/.julia/packages/MbedTLS/a1JFn/deps/usr/lib/libmbedtls.2.16.0.dylib (unknown line)
mbedtls_ssl_handshake at /Users/kristoffer/.julia/packages/MbedTLS/a1JFn/deps/usr/lib/libmbedtls.2.16.0.dylib (unknown line)
ssl_handshake at /Users/kristoffer/.julia/packages/MbedTLS/a1JFn/src/ssl.jl:606
handshake at /Users/kristoffer/.julia/packages/MbedTLS/a1JFn/src/ssl.jl:74
jl_apply at /Users/sabae/buildbot/worker/package_macos64/build/src/./julia.h:1692 [inlined]
do_call at /Users/sabae/buildbot/worker/package_macos64/build/src/interpreter.c:369
eval_body at /Users/sabae/buildbot/worker/package_macos64/build/src/interpreter.c:0
jl_interpret_toplevel_thunk at /Users/sabae/buildbot/worker/package_macos64/build/src/interpreter.c:911
top-level scope at /Users/kristoffer/.julia/packages/MbedTLS/a1JFn/test/runtests.jl:123
jl_toplevel_eval_flex at /Users/sabae/buildbot/worker/package_macos64/build/src/toplevel.c:814
jl_parse_eval_all at /Users/sabae/buildbot/worker/package_macos64/build/src/ast.c:872
jl_load at /Users/sabae/buildbot/worker/package_macos64/build/src/toplevel.c:872 [inlined]
jl_load_ at /Users/sabae/buildbot/worker/package_macos64/build/src/toplevel.c:879
include at ./client.jl:439
jl_apply at /Users/sabae/buildbot/worker/package_macos64/build/src/./julia.h:1692 [inlined]
do_call at /Users/sabae/buildbot/worker/package_macos64/build/src/interpreter.c:369
eval_body at /Users/sabae/buildbot/worker/package_macos64/build/src/interpreter.c:0
jl_interpret_toplevel_thunk at /Users/sabae/buildbot/worker/package_macos64/build/src/interpreter.c:911
top-level scope at none:6
jl_toplevel_eval_flex at /Users/sabae/buildbot/worker/package_macos64/build/src/toplevel.c:814
jl_toplevel_eval_flex at /Users/sabae/buildbot/worker/package_macos64/build/src/toplevel.c:764
jl_toplevel_eval at /Users/sabae/buildbot/worker/package_macos64/build/src/toplevel.c:823 [inlined]
jl_toplevel_eval_in at /Users/sabae/buildbot/worker/package_macos64/build/src/toplevel.c:843
eval at ./boot.jl:331
exec_options at ./client.jl:264
_start at ./client.jl:484
jfptr__start_2076.clone_1 at /Applications/Julia-1.5.app/Contents/Resources/julia/lib/julia/sys.dylib (unknown line)
true_main at /Applications/Julia-1.5.app/Contents/Resources/julia/bin/julia (unknown line)
main at /Applications/Julia-1.5.app/Contents/Resources/julia/bin/julia (unknown line)

signal (11): Segmentation fault: 11
in expression starting at /Users/kristoffer/.julia/packages/MbedTLS/a1JFn/test/runtests.jl:89
jl_field_offset at /Users/sabae/buildbot/worker/package_macos64/build/src/./julia.h:1018 [inlined]
set_nth_field at /Users/sabae/buildbot/worker/package_macos64/build/src/datatype.c:1085
jl_f_tuple at /Users/sabae/buildbot/worker/package_macos64/build/src/builtins.c:725
jl_method_error at /Users/sabae/buildbot/worker/package_macos64/build/src/gf.c:1734
jl_lookup_generic_ at /Users/sabae/buildbot/worker/package_macos64/build/src/gf.c:2297
jl_apply_generic at /Users/sabae/buildbot/worker/package_macos64/build/src/gf.c:2318
f_rng at /Users/kristoffer/.julia/packages/MbedTLS/a1JFn/src/ctr_drbg.jl:22
unknown function (ip: 0x11f36af7c)
ssl_generate_random at /Users/kristoffer/.julia/packages/MbedTLS/a1JFn/deps/usr/lib/libmbedtls.2.16.0.dylib (unknown line)
ssl_write_client_hello at /Users/kristoffer/.julia/packages/MbedTLS/a1JFn/deps/usr/lib/libmbedtls.2.16.0.dylib (unknown line)
mbedtls_ssl_handshake_client_step at /Users/kristoffer/.julia/packages/MbedTLS/a1JFn/deps/usr/lib/libmbedtls.2.16.0.dylib (unknown line)
mbedtls_ssl_handshake_step at /Users/kristoffer/.julia/packages/MbedTLS/a1JFn/deps/usr/lib/libmbedtls.2.16.0.dylib (unknown line)
mbedtls_ssl_handshake at /Users/kristoffer/.julia/packages/MbedTLS/a1JFn/deps/usr/lib/libmbedtls.2.16.0.dylib (unknown line)
ssl_handshake at /Users/kristoffer/.julia/packages/MbedTLS/a1JFn/src/ssl.jl:606
handshake at /Users/kristoffer/.julia/packages/MbedTLS/a1JFn/src/ssl.jl:74
jl_apply at /Users/sabae/buildbot/worker/package_macos64/build/src/./julia.h:1692 [inlined]
do_call at /Users/sabae/buildbot/worker/package_macos64/build/src/interpreter.c:369
eval_body at /Users/sabae/buildbot/worker/package_macos64/build/src/interpreter.c:0
jl_interpret_toplevel_thunk at /Users/sabae/buildbot/worker/package_macos64/build/src/interpreter.c:911
jl_toplevel_eval_flex at /Users/sabae/buildbot/worker/package_macos64/build/src/toplevel.c:814
jl_parse_eval_all at /Users/sabae/buildbot/worker/package_macos64/build/src/ast.c:872
jl_load at /Users/sabae/buildbot/worker/package_macos64/build/src/toplevel.c:872 [inlined]
jl_load_ at /Users/sabae/buildbot/worker/package_macos64/build/src/toplevel.c:879
include at ./client.jl:439
jl_apply at /Users/sabae/buildbot/worker/package_macos64/build/src/./julia.h:1692 [inlined]
do_call at /Users/sabae/buildbot/worker/package_macos64/build/src/interpreter.c:369
eval_body at /Users/sabae/buildbot/worker/package_macos64/build/src/interpreter.c:0
jl_interpret_toplevel_thunk at /Users/sabae/buildbot/worker/package_macos64/build/src/interpreter.c:911
jl_toplevel_eval_flex at /Users/sabae/buildbot/worker/package_macos64/build/src/toplevel.c:814
jl_toplevel_eval_flex at /Users/sabae/buildbot/worker/package_macos64/build/src/toplevel.c:764
jl_toplevel_eval at /Users/sabae/buildbot/worker/package_macos64/build/src/toplevel.c:823 [inlined]
jl_toplevel_eval_in at /Users/sabae/buildbot/worker/package_macos64/build/src/toplevel.c:843
eval at ./boot.jl:331
exec_options at ./client.jl:264
_start at ./client.jl:484
jfptr__start_2076.clone_1 at /Applications/Julia-1.5.app/Contents/Resources/julia/lib/julia/sys.dylib (unknown line)
true_main at /Applications/Julia-1.5.app/Contents/Resources/julia/bin/julia (unknown line)
main at /Applications/Julia-1.5.app/Contents/Resources/julia/bin/julia (unknown line)
Allocations: 3464917 (Pool: 3464201; Big: 716); GC: 4
```

</details>

KristofferC commented:
Regarding cbrt getting 11x slower (and cos, sin etc) I think this is just benchmark noise.

On 1.3:

```
julia> @btime cbrt($(nextfloat(zero(Float64))));
  0.029 ns (0 allocations: 0 bytes)
```
and master

```
julia> @btime cbrt($(nextfloat(zero(Float64))));
  155.990 ns (0 allocations: 0 bytes)
```

Doing the "Ref trick":

1.3:

```
julia> @btime cbrt($(Ref(nextfloat(zero(Float64))))[]);
  178.467 ns (0 allocations: 0 bytes)
```

and master

```
julia> @btime cbrt($(Ref(nextfloat(zero(Float64))))[]);
  156.010 ns (0 allocations: 0 bytes)
```

KristofferC commented:
Allocation increase in `["array", "index", "(\"sumvector\", \"1.0:0.00010001000100010001:2.0\")"]` is MWEd with

```
julia> v = 1.0:0.00010001000100010001:2.0
1.0:0.00010001000100010001:2.0

julia> r = rand(1:length(v), 5);

julia> f(v, r) = v[r, 1]
f (generic function with 1 method)

julia> @btime f($v, $r)
  76.075 ns (3 allocations: 192 bytes)
```

(makes 2 allocations on 1.3).

sostock commented:
Before releasing a new HalfIntegers version that is compatible with Julia 1.4 (and marking old versions as incompatible), I wanted to know whether #34129 could be included in 1.4 as well. I guess it’s too late for that now?

While it is not a bugfix, I think it is an important change to the implementation in #33910 (which was merged shortly before the 1.4 branch happened). Without #34129, it is rather inconvenient to extend `gcd`/`lcm`/`gcdx` to custom `Real` types (see https://github.com/JuliaLang/julia/pull/33910#issuecomment-566182509).

JeffBezanson committed:
remove code forcing testsets to be compiled (#34405)

(cherry picked from commit ed87ece188bea86893623fa25fca7005033e681f)

sostock committed:
Fix array-argument gcd for non-integers

(cherry picked from commit 3549cd190ecd0574c803de80e5f6d8d70cbf05e3)

musm committed:
Use UUID's to generate random tempname on Windows (#33785)

* Use UIUD to create random tempname on Windows

* Use underscores and remove extension

* Truncate to 10 chars the UUID

* Generate the random name from a random byte array

* Update file.jl

(cherry picked from commit d759b5b3343d1b2654453b06f7e802a76c9e27e6)

Keno committed:
Check return value of SystemFunction036 (#34429)

In both places we call it. Also tabs-to-spaces for the new code
to match the formatting of the rest of the code.

(cherry picked from commit f120989cbdee2cc075ad52388cda2959aeb33e9d)

JeffBezanson committed:
fix #34415, remove affinity setting code (#34433)

This was removed before and added back apparently by mistake.

(cherry picked from commit 592878623d376c71e5452dc2775fa2f7a4e097ca)

KristofferC commented:
@sostock, your changes should be in this branch now.
